### PR TITLE
[special] add zeta

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #50345

`zeta` was already present in the codebase to support computation of `polygamma`.

However, `zeta` only had `double(double, double)` signature **for CPU** before the PR (which meant that computation `polygamma` were always upcasted to `double` for zeta part).

With this PR, float computations will take place in float and double in double.

Have also refactored the code and moved the duplicate code from `Math.cuh` to `Math.h`

**Note**: For scipy, q is optional, and if it is `None`, it defaults `1` which corresponds to Reimann-Zeta. However, for `torch.specia.zeta`, I made it mandatory cause for me it feels odd without `q` this is Reimann-Zeta and with `q` it is the general Hurwitz Zeta. I think sticking to just general made more sense as passing `1` for q sounds trivial.

Verify:
* [x] Docs https://14234587-65600975-gh.circle-artifacts.com/0/docs/special.html#torch.special.zeta

### CODE REVIEW

File: test/test_binary_ufuncs.py [hunk 1]
```diff
@@ -2761,6 +2761,60 @@ def _compare_helper(x, y, torch_fn, reference_fn):
...
+        def test_helper(x, q):
+            x_np = x if isinstance(x, float) else x.cpu().numpy()
+            q_np = q if isinstance(q, float) else q.cpu().numpy()
+            expected = torch.from_numpy(scipy.special.zeta(x_np, q_np))
+            actual = torch.special.zeta(x, q)
+
+            rtol, atol = None, None
+            if self.device_type == 'cpu':
+                rtol, atol = 1e-6, 1e-6
+            self.assertEqual(expected, actual, rtol=rtol, atol=atol, exact_dtype=False)
```
Comment (kshitij12345) [inline test/test_binary_ufuncs.py:2781]: Had to add tolerance based on CI failures (for jobs with relatively old compilers).

https://app.circleci.com/pipelines/github/pytorch/pytorch/338469/workflows/db193abd-5b89-405a-9987-f4926107e2ec/jobs/14225591

https://app.circleci.com/pipelines/github/pytorch/pytorch/338469/workflows/db193abd-5b89-405a-9987-f4926107e2ec/jobs/14224886

Comment (mruberry) [reply test/test_binary_ufuncs.py:2781]: cc @pmeier for yet another example of binary elementwise testing we should automate

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -6972,6 +6985,25 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            safe_casts_outputs=True,
            supports_forward_ad=True,
            sample_inputs_func=sample_inputs_xlog1py),
+    OpInfo('special.zeta',
+           aten_name='special_zeta',
+           dtypes=all_types_and(torch.bool),
+           supports_autograd=False,
+           safe_casts_outputs=True,
+           sample_inputs_func=sample_inputs_binary_pwise),
+    # OpInfo entry to verify the gradient formula of `other`/`q`
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6994]: Smart.

One goal of our "writing ops in Python" project (which we're looking at for H2 2021) is to simplify supporting scalar x tensor and tensor x scalar operations, possibly by automating the generation of those functions

**Review by mruberry — APPROVED** [review]
Another great, thorough, careful PR @kshitij12345!

**Comment by mruberry** [comment]
@kshitij12345 would you just merge this and ping me?

**Comment by kshitij12345** [comment]
@mruberry Done :)

**Pull request closed without merging**